### DISCUSSION TITLE
Modernizing the avatar controls

### DISCUSSION THREAD

**Discussion by dgelessus** [H-uru/Plasma#1203] [General]
A semi-common request from players is to update the avatar controls to make them work more like modern games, compared to Uru's current "tank controls" that are a bit dated (and depending on who you ask, just not very good).

@dpogue asked for someone to write up what, specifically, would need to be changed about the controls to modernize them - so here is what I would have in mind. Everyone is of course welcome to add their own opinions and suggestions!

The short version of this proposal is: let Uru behave like Myst V's "Free-Move" mode and the realMyst games. I'll focus on first-person mode first.

### Mouse look

This would be based on Uru's current first-person controls, with the following changes:

* The cursor is locked to the center of the screen by default. Moving the mouse rotates the camera, like when holding secondary mouse with the current first-person controls.
* A secondary mouse click unlocks the cursor, allowing it to move freely across the screen without moving the camera. When secondary mouse is clicked a second time, or when the avatar is moved, then the mouse is locked to the camera again.
  * This mode is enabled automatically while interacting with GUIs that lock the camera: popups (e. g. dialogs, big KI, books) and in-world control panels (e. g. Teledahn control room panel, Nexus book machine).
* Other mouse controls behave like they currently do in first-person mode. Primary mouse interacts with hotspots and otherwise walks forward when held. Holding secondary mouse together with primary mouse toggles running. Holding mouse 3 walks backwards.

### WASD movement keys

You can already get limited WASD controls in Uru by customizing the key map: W/S for forward/backward, A/D for step left/right, and Q/E for rotate left/right. But fundamentally that still only gives you tank controls, just on different keys. "True" modern WASD controls would behave differently in a few ways:

* Sideways movement (A/D) is much faster than Uru's step left/right controls - probably about as fast as walking forward. Most games call this "strafe" and not "side-step" like Uru.
  * In games where you can see your own body, the corresponding animation is a "sideways walk": the avatar turns slightly towards the direction to which it's strafing and walks by moving one foot over/past the other. This is unlike Uru's side-step animation, where the avatar stays looking straight ahead and shuffles sideways, bumping its feet against each other.
  * I am aware that changes to avatar animations are very unlikely to happen at this point, and new animations even less so. Someone on Discord suggested "just" speeding up the side-step animation. I'm guessing that won't be super easy either, but probably more feasible than creating a new animation from scratch.
* It's possible to walk forward/backward (W/S) and strafe (A/D) at the same time, which results in walking diagonally.
  * I have no idea how feasible this is animation-wise. Worst-case, perhaps use the forward walk/run animation, possibly at an angle?
* It's possible to jump not just while walking straight ahead, but also while walking diagonally forward, and possibly also when fully strafing left/right.
  * This one is less important than the previous points. Again, I'm guessing this will be difficult animation-wise.
* Combined with mouse look as described above, there is no need for rotate left/right keys anymore, so Q/E can be freed up.
* This would basically require setting an explicit chat start key, because Uru's implicit chat won't work well with four common letters occupied by movement keys. Common chat start keys in other games are T and C - though personally I'm a fan of Enter (for symmetry with sending a chat message).

### Third-person mode

This is a bit more difficult. I'm less familiar with games where you control an avatar in third person, but I think the common control schemes in such games are:

* Camera locked straight behind the avatar. Moving the mouse immediately changes the orientation of the avatar. This is basically like first-person mode with the camera shifted 1-2 meters back, or like Uru's default third-person camera combined with mouse look. Example: Minecraft's third-person mode (F5).
* Camera orbits the avatar. Similar to the first variant, but allows looking at your avatar from any angle. Avatar movement is relative to the camera orientation - e. g. W always walks away from the camera, with the avatar rotating automatically once you start walking. Example: Mass Effect (I think? Never actually played it myself.)
* Camera locked at a fixed angle, looking down at the avatar. The 3D equivalent of isometric games. Avatar movement is relative to the camera - e. g. W always walks "up". Example: Magicka.

The third mode wouldn't work in Uru - in almost all ages, there's no single fixed camera angle that would allow you to see everything properly, at least not without significant changes like conditionally hiding parts of the geometry.

The other two modes would both work in principle, but there are a few problems:

* Uru often uses a custom third-person camera placement with a different angle than the default. Because avatar movement in these modes is relative to the camera, it would be impossible to control the avatar across forced camera changes.
  * The obvious solution would be to disable all automatic camera changes, except during cutscenes where the player can't move the avatar. I'm not sure if this would cause problems anywhere though.
* The cursor would be locked on top of the avatar, making it near-impossible to hit any hotspots. Uru's current third-person mode solves this by making the avatar transparent when you mouse over it, but with the mouse always over the avatar, that quickly gets silly (and spooky 👻).
  * This could be worked around using an over-the-shoulder camera, where the locked cursor always points past the avatar
  * We should also allow unlocking the cursor from the camera, like in first-person mouse look mode, to allow interacting with oddly placed hotspots without having to move the camera and/or avatar too wildly.

### Finally...

I think this is relatively obvious: any changes to the controls should be *optional*, allowing players to choose whichever control scheme they prefer. The original controls have been around for 20 years, so people *will* be unhappy if you take them away.

A different question is whether the "modern" controls should be the default for new players. I have no strong opinions about this right now, and I think it's better answered after the new controls in question have been implemented and tested for a bit.

**Comment by jfmherokiller** [comment]
for me atleast i would look forward to better wasd controls because outside of uru thats the standard

**Comment by patmauro** [comment]
Wanted to chime in here - having troubleshooted with several new players over time, I've often found that the discomfort new players seem to have with the controls stems _less_ from the tank controls _themselves_ - which the game is largely built around, and folks usually get the hang of pretty quickly - and more the _literal_ arrow keys, which newer players simply find completely alien as input. Newer players (a) seem to largely be used to using "WASD" for directional movement, and (b) find the arrow keys themselves to be very uncomfortable use.

In my experience with new players, this is largely solved by mapping WASD as alternates to the directional keys & mapping "start chat" to Enter:
![image](https://github.com/H-uru/Plasma/assets/6642931/28803919-9713-4d02-bdc7-f3e7d5600d3b)

**I _don't_ advocate overwriting the default with this**, as some returning users would certainly take issue with this. However: offering this as an _option_ to _new_ players on first startup would be a really great way of handling this - similar to the prompt Myst V shows at startup:
![image](https://github.com/H-uru/Plasma/assets/6642931/d6c8631c-1ad3-450b-8a97-c9c247436efd)

So - while I'm definitely still in favor of the much more radical experiments proposed above, as I'm sure they will bear a lot of interesting fruit, I'd like to offer up this as an interim suggestion - something that is actionable right now, should be quick, and would give us a solid accessibility W tackling a frequently asked question.

**Reply by dpogue** [reply]
> **I _don't_ advocate overwriting the default with this**, as some returning users would certainly take issue with this. However: offering this as an _option_ to _new_ players on first startup would be a really great way of handling this - similar to the prompt Myst V shows at startup:
> ![image](https://private-user-images.githubusercontent.com/6642931/311074459-d6c8631c-1ad3-450b-8a97-c9c247436efd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk4NTcyODMsIm5iZiI6MTcwOTg1Njk4MywicGF0aCI6Ii82NjQyOTMxLzMxMTA3NDQ1OS1kNmM4NjMxYy0xYWQzLTQ1MGItOGE5Ny1jOWMyNDc0MzZlZmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDhUMDAxNjIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDIyNjk2ZGFlYTY4ODY1ODU0YzY3NzYyZWRkMmE4ZDFmMTJiMjY0MmNiNTgwYjE4NzYyNDY2Y2YyMGNjMDhkMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.rkcAas4-qr0HXU_WYePKH3Ui9kYyGz8tqTfYLN8NR9w)

There is a popup when first starting a new player for "basic" or "advanced" mouse settings IIRC, something like this seems like a good addition (although I would prefer to just say "arrow keys" or "WASD" without labelling either as basic/advanced. I don't think WASD is more advanced than arrow keys, it's just different keys)

One thing to be cautious of is international keyboard mappings (QWERTY vs AZERTY, etc.)

**Reply by patmauro** [reply]
Yes, that's right. Perhaps expanding this menu option would be the perfect place for it?
![image](https://github.com/H-uru/Plasma/assets/6642931/a4aea4de-3def-4848-934b-6d665ad5b48c)

**Reply by patmauro** [reply]
The complaints from newbies continue:
![image](https://github.com/user-attachments/assets/86a76d79-3e48-4c6a-89d8-7b686114404f)

Obviously there's a lot going on here, but there's clearly a strong need for us to make things a little more user friendly. As discussed before, while I understand many folks are uncomfortable with the idea of overwriting the default controls, as a compromise, I'd still like to once again advocate for adding an option to set this automatically to the startup prompt, alongside the mouse presets:
![image](https://github.com/user-attachments/assets/941e6089-c33c-4af7-90a4-a36080495f53)

I also think we may want to consider changing the language for F1 on this screen from "Switch Views" to the more commonly-used "Toggle First-Person/Third-Person"
![image](https://github.com/user-attachments/assets/53da37f7-3e5c-4b45-a3a8-3b42eb9571b3)

**Reply by Deledrius** [reply]
That's useful feedback, although I'm at a loss to imagine how they could ever find it impossible to walk in a straight line in Uru.  That's the one thing it does well, control-wise.  It's hard to say what was going on here, but it's clear that modernizing the controls would go a long way to prevent it from happening.

**Comment by Mystler** [comment]
I'm back playing some URU after a while as I revisit the Myst series and, while URU's controls are perfectly usable, I find myself also missing modern WASD first-person controls. Making it super extravagant with new animations is probably not realistic but if the direction of the first person camera can be untangled from the animation that is being played, then that sounds like it might already do the trick? Strictly speaking first person here. I understand third is a different beast, but probably fine as is.

W+A for example would not turn left as it does now but just play the normal forward running animation at a 45° angle to the left of the camera's forward direction. For backwards it could likewise just do that in a backwards direction but if gaining massive speed from that is an actual concern it could still play the normal backwards animation and do the same thing.

Oh, and while at it. Binding W to forward already works but i'm not a fan of having to use caps lock to sprint. I would love a toggle instead of hold to sprint mode (and/or potentially a setting to invert the default behavior).

**Reply by TheScarFr** [reply]
Currently, to run, you can maintain the Shift key + the top arrow or a custom key set to move forward or the mouse's left click, you can also toggle the run by pressing the Caps Lock key, and you can also run by pressing the mouse's left+right clicks.

**Reply by Mystler** [reply]
Yes, I am aware. I'm not sure i follow what point you are trying to make.

**Reply by TheScarFr** [reply]
> Binding W to forward already works but i'm not a fan of having to use caps lock to sprint. I would love a toggle instead of hold to sprint mode (and/or potentially a setting to invert the default behavior).

You asked for a toggle of the run (and/or potentially a setting to invert the default behavior).
The existing toggle currently works for 1) the top arrow key, 2) the mouse's left click, and I think also for 3) a custom key set to move forward; so, what you ask for is already there. Maybe I'm wrong about the 3rd point.
Regarding the parenthesis, yes, this function is missing.

**Reply by Mystler** [reply]
As you have literally quoted from myself
> i'm not a fan of having to use caps lock to sprint

So it should be clear i am aware caps lock toggles but, since this is a *suggestion* thread for _modernizing_ controls, I am making the *suggestion* and opening discussion on the fact that Caps lock is very suboptimal as a sprint toggle when it also affects typing on a system wide level. I have made the attempt to point out that many modern games have a "toggle vs hold" mode setting that applies to the regular sprint button (usually Shift) and that something like this would be a preferred solution to me.

Also this was just a side note suggestion as part of a larger post.

**Reply by TheScarFr** [reply]
Ok, you were not clear enough to me...
Your last message fixed that.
